src/utils/refreshDockerode.ts

```diff
@@ -0,0 +1,104 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { DockerOptions } from 'dockerode';
+import Dockerode = require('dockerode');
+import { ext } from '../extensionVariables';
+import { addDockerSettingsToEnv } from './addDockerSettingsToEnv';
+import { cloneObject } from './cloneObject';
+import { execAsync } from './execAsync';
+import { isWindows } from './osUtils';
+
+const unix = 'unix://';
+const npipe = 'npipe://';
+
+const SSH_URL_REGEX = /ssh:\/\//i;
+
+// Not exhaustive--only the properties we're interested in
+interface IDockerEndpoint {
+    Host?: string;
+}
+
+// Also not exhaustive--only the properties we're interested in
+interface IDockerContext {
+    Endpoints: { [key: string]: IDockerEndpoint }
+}
+
+/**
+ * Dockerode parses and handles the well-known `DOCKER_*` environment variables, but it doesn't let us pass those values as-is to the constructor
+ * Thus we will temporarily update `process.env` and pass nothing to the constructor
+ */
+export async function refreshDockerode(): Promise<void> {
+    const oldEnv = process.env;
+    try {
+        const newEnv: NodeJS.ProcessEnv = cloneObject(process.env); // make a clone before we change anything
+        addDockerSettingsToEnv(newEnv, oldEnv);
+
+        const dockerodeOptions = await getDockerodeOptions(newEnv);
+
+        ext.dockerodeInitError = undefined;
+        process.env = newEnv;
+        ext.dockerode = new Dockerode(dockerodeOptions);

```

Comment:
Nit: the `finally` need only follow this statement (i.e. only reset `process.env` if it was actually changed) and is independent of the `catch`.

Reply:
@philliphoff At a minimum, `getDockerodeOptions` must be in the `try` block, and `addDockerSettingsToEnv` should probably be, so we can't shrink the try block more than a line. How would you structure the `finally` to not change `process.env` unless needed, without making e.g. a boolean to keep track of the fact that we changed it?

Reply:
It'd require two `try` blocks, one for the `catch` and the other for the `finally`.  Something like:

```typescript
const oldEnv = process.env;
try {
  ...
  process.env = newEnv;
  try {
    ext.dockerode = new Dockerode(dockerodeOptions);
  } finally {
    process.env = oldEnv;
  }
}
catch (error) {
  ...
}
```